utils/src/main/java/software/amazon/awssdk/utils/async/FlatteningSubscriber.java

```diff
@@ -76,25 +86,38 @@ public void onNext(Iterable<U> nextItems) {
     }
 
     private void fulfillDemand() {
+        if (terminalCallMadeDownstream) {
+            return;
+        }
+
         while (demand.get() > 0 && !currentBatch.isEmpty()) {
             demand.decrementAndGet();
             subscriber.onNext(currentBatch.poll());
         }
 
-        if (onCompleteCalled && currentBatch.isEmpty()) {
+        if (onCompleteCalledByUpstream && currentBatch.isEmpty()) {
             subscriber.onComplete();
         } else if (currentBatch.isEmpty() && demand.get() > 0) {
             requestedNextBatch = true;
             sourceSubscription.request(1);
         }
     }
 
+    @Override
+    public void onError(Throwable throwable) {
+        synchronized (lock) {
+            subscriber.onError(throwable);
+            terminalCallMadeDownstream = true;
```

Comment:
Sure, wouldn't hurt.

Reply:
It's tricky to do this safely, and I don't see much real benefit. Is there a specific reason you think we should do it?

Reply:
To prevent out of memory error for slow-yet-demanding subscriber after the subscription is cancelled or there's an error I guess. I don't think this is a blocker for this PR.

Reply:
This never holds more than one "batch" of items in memory at once. It only requests more from upstream 1 at a time, no matter how many the downstream asks for.

Reply:
Oh I see, then I agree there is not much benefit